File: ai-backend/src/services/charter_playground_chat/charter_playground_chat_schema.py
from typing import List, Literal, Optional

from pydantic import BaseModel

from src.internal_services.llm_client.llm_client_base import LLMClientName, LLMModelType
from src.services.charter_metric.charter_metrics_schema import CharterMetricResponse
from src.services.charter_metric_example.charter_metric_example_schema import CharterMetricExampleResponse
from src.services.data_entity.data_entity_schema import DataEntityResponse


class PlaygroundMessageContext(BaseModel):
    data_entity_ids: Optional[List[int]] = None
    charter_metric_ids: Optional[List[int]] = None

    selected_texts: Optional[List[str]] = None
    query: Optional[str] = None

    data_entities: Optional[List[DataEntityResponse]] = None
    charter_metrics: Optional[List[CharterMetricResponse]] = None
    charter_metric_examples: Optional[List[CharterMetricExampleResponse]] = None


class PlaygroundMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str
    context: Optional[PlaygroundMessageContext] = None


class PlaygroundChatInput(BaseModel):
    messages: List[PlaygroundMessage]
    model_type: Optional[LLMModelType] = None
    client_name: Optional[LLMClientName] = None
    conversation_id: Optional[str] = None


class PlaygroundChatResponse(PlaygroundChatInput):
    pass

File: analyst-ui/services/charterPlaygroundChat/charterPlaygroundChat.schema.ts
import { z } from "zod";
import { DataEntityResponseSchema } from "../dataEntity/dataEntity.schema";
import { CharterMetricSchema } from "../charterMetric/charterMetric.schema";
import { CharterMetricExampleSchema } from "../charterMetricExample/charterMetricExample.schema";

export type ModelNames = "gpt-4o" | "gpt-4o-mini" | "claude-3.5-haiku" | "claude-3.5-sonnet";

// Mapping of model names to model types and client names
export const ModelNameMapping: Record<ModelNames, { model_type: "large" | "small"; client_name: "openai" | "claude" }> =
  {
    "gpt-4o": { model_type: "large", client_name: "openai" },
    "gpt-4o-mini": { model_type: "small", client_name: "openai" },
    "claude-3.5-haiku": { model_type: "small", client_name: "claude" },
    "claude-3.5-sonnet": { model_type: "large", client_name: "claude" },
  };

// Importing related schemas (you'll need to adjust the import paths)
export const PlaygroundMessageContextSchema = z.object({
  query: z.string().optional(),
  data_entity_ids: z.array(z.number()).optional(),
  charter_metric_ids: z.array(z.number()).optional(),
  selected_texts: z.array(z.string()).optional(),

  data_entities: z.array(DataEntityResponseSchema).optional(),
  charter_metrics: z.array(CharterMetricSchema).optional(),
  charter_metric_examples: z.array(CharterMetricExampleSchema).optional(),
});

export const PlaygroundMessageSchema = z.object({
  role: z.enum(["user", "assistant"]),
  content: z.string(),
  context: PlaygroundMessageContextSchema.optional().nullable(),
});

export const PlaygroundChatInputSchema = z.object({
  messages: z.array(PlaygroundMessageSchema),
  model_type: z.enum(["small", "large"]),
  client_name: z.enum(["openai", "claude"]),
  conversation_id: z.string().optional().nullable(),
});

// Type definitions for TypeScript
export type PlaygroundMessageContext = z.infer<typeof PlaygroundMessageContextSchema>;
export type PlaygroundMessage = z.infer<typeof PlaygroundMessageSchema>;
export type PlaygroundChatInput = z.infer<typeof PlaygroundChatInputSchema>;
